HPCC-21121 Maintain XSD attribute and element order as set in XSDs

kenrowland opened the pull request:
Add ordinal value to attribute for use by UI
Store elements in vector to maintain order

Signed-off-by: Ken Rowland <kenneth.rowland@lexisnexisrisk.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [x] Security
  - [x] Thread-safety
  - [x] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

kenrowland commented:
@rpastrana Please review

HPCCSmoketest commented:
Process of PR-11965, label: HPCC-21121 starts now. Estimated completion time is ~ 0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: cbc78f70ee99da11a7031f90eb021b4be23e0862
Build: success
Build: success
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |95 |95 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |810 |810 |0 |
|test (thor) |734 |734 |0 |
|test (roxie) |883 |883 |0 |


HPCC Stop: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   26 sec (00:00:26) |  188 sec (00:03:08) |    0 sec (00:00:00) |    2 sec (00:00:02) |   19 sec (00:00:19) | 1478 sec (00:24:38) |   18 sec (00:00:18) | 1731 sec (00:28:51) |

rpastrana commented on configuration/configmgr/configmgrlib/SchemaValue.cpp:
```diff
@@ -32,6 +32,7 @@ SchemaValue::SchemaValue(const std::string &name, bool isDefined) :
     bitMask.m_deprecated = 0;
     bitMask.m_isUnique = 0;
     bitMask.m_isDefined = isDefined;
+    m_ordinal = 0;
```
constructor initializer list?

kenrowland replied:
Since this is a native object, direct initialization vs an initializer list is equivalent. There is no construction so direct assignment is OK. If there is a desire to have it in the initializer list, I will move it.

rpastrana commented on configuration/configmgr/configmgrlib/SchemaValue.cpp:
```diff
@@ -59,6 +60,8 @@ SchemaValue::SchemaValue(const SchemaValue &value)
     std::vector<std::shared_ptr<SchemaValue>> m_mirrorToSchemaValues;
     std::vector<std::weak_ptr<SchemaValue>> m_pUniqueValueSetRefs;    // this value serves as the key from which values are valid
+
+    m_ordinal = 0;  // assigned when inserted
```
constructor initializer list?

kenrowland replied:
The constructor for a SchemaValue will set it to 0, however, should the code in the copy constructor, where this line is, accidentally copy the value from the source object, this will make sure it is 0 when the copy is complete. I felt it just safer to explicitly set it to 0 before exit.

rpastrana reviewed (changes requested):
@kenrowland a few minor comments

kenrowland commented on configuration/configmgr/configmgrlib/SchemaItem.hpp:
```diff
@@ -98,7 +97,8 @@ class DECL_EXPORT SchemaItem : public std::enable_shared_from_this<SchemaItem>
         unsigned m_maxInstances;
-        std::multimap<std::string, std::shared_ptr<SchemaItem>> m_children;
+        std::vector<std::shared_ptr<SchemaItem>> m_children;
+//        std::multimap<std::string, std::shared_ptr<SchemaItem>> m_children;
```
Removed.

kenrowland reviewed (commented):
@rpastrana Please see changes and responses.

HPCCSmoketest commented:
Process of PR-11965, label: HPCC-21121 starts now. Estimated completion time is ~ 0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 834451a68326352fbc01bb289271ddbce1fa6696
Build: success
Build: success
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |95 |95 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |810 |810 |0 |
|test (thor) |734 |734 |0 |
|test (roxie) |883 |883 |0 |


HPCC Stop: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   24 sec (00:00:24) |  185 sec (00:03:05) |    0 sec (00:00:00) |    2 sec (00:00:02) |   18 sec (00:00:18) | 1475 sec (00:24:35) |   19 sec (00:00:19) | 1723 sec (00:28:43) |

kenrowland commented:
@rpastrana Not sure if I pushed this back to you or not, but can you please see my responses.

HPCCSmoketest commented:
Process of PR-11965, label: HPCC-21121 starts now. Estimated completion time is ~ 0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :x:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 834451a68326352fbc01bb289271ddbce1fa6696
Number of Error(s): 111
/mnt/disk1/home/vamosax/smoketest/PR-11965/HPCC-Platform/configuration/configmgr/configmgrlib/SchemaValue.hpp:95:1: error: expected unqualified-id before \xe2\x80\x98<<\xe2\x80\x99 token

[Build error:](https://gist.github.com/832cc49efae974a955953653569a6814)
Build log [Build log](https://gist.github.com/HPCCSmoketest/832cc49efae974a955953653569a6814#file-result-18-12-10-15-15-27-log)

Commit:
HPCC-21121 Maintain XSD attribute and element order as set in XSDs

Add ordinal value to attribute for use by UI
Store elements in vector to maintain order

Signed-off-by: Ken Rowland <kenneth.rowland@lexisnexisrisk.com>

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 3c181e31cff4970ab82aa2a8665d660f524b825c
Build: success
Build: success
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |99 |99 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |812 |812 |0 |
|test (thor) |738 |738 |0 |
|test (roxie) |885 |885 |0 |


HPCC Stop: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   29 sec (00:00:29) |  189 sec (00:03:09) |    0 sec (00:00:00) |    2 sec (00:00:02) |   18 sec (00:00:18) | 1456 sec (00:24:16) |   20 sec (00:00:20) | 1714 sec (00:28:34) |

rpastrana reviewed (approved):
@kenrowland looks reasonable

kenrowland commented:
@richardkchapman Please merge
